File: src/cli/paratoric.cpp
// ParaToric - Continuous-time QMC for the extended toric code in the x/z-basis
// Copyright (C) 2022-2026  Simon Mathias Linsel, Lode Pollet

#include "io/io.hpp"
#include "mcmc/input_validation.hpp"
#include "paratoric/types/types.hpp"

#include <boost/log/core.hpp> 
#include <boost/log/expressions.hpp>
#include <boost/log/trivial.hpp> 
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/value_semantic.hpp>
#include <boost/program_options/variables_map.hpp>

#include <chrono>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <string_view>
#include <vector>

namespace po = boost::program_options;

int main(int argc, char **argv) {
    try {
        paratoric::Config config{};
        std::string simulation;
        int process_index = 0;

        po::options_description desc("Run extended toric code QMC simulation");
        desc.add_options()
        ("help,h", "Display this help message.")
        ("simulation,sim", po::value(&simulation), "Simulation you want to run.")
        ("N_samples,Ns", po::value(&config.sim_spec.N_samples), "Number of samples taken during the simulation.")
        ("N_thermalization,Nth", po::value(&config.sim_spec.N_thermalization), "Number of thermalization steps.")
        ("N_between_samples,Nbs", po::value(&config.sim_spec.N_between_samples), "Number of steps between samples.")
        ("beta,bet", po::value(&config.lat_spec.beta), "Inverse temperature.")
        ("h_constant,hc", po::value(&config.param_spec.h), "Value of the constant h (electric field term).")
        ("h_hysteresis,hhys", po::value(&config.param_spec.h_hys)->multitoken(), "Hysteresis values of the constant h (electric field term).")
        ("h_constant_therm,hct", po::value(&config.param_spec.h_therm), "Value of the constant h for thermalization (electric field term).")
        ("mu_constant,muc", po::value(&config.param_spec.mu), "Value of the constant mu_constant (star term).")
        ("J_constant,Jc", po::value(&config.param_spec.J), "Value of the constant J (plaquette term).")
        ("lmbda_constant,lmbdac", po::value(&config.param_spec.lmbda), "Value of the constant lmbda (gauge field term).")
        ("lmbda_hysteresis,lmbdahys", po::value(&config.param_spec.lmbda_hys)->multitoken(), "Hysteresis values of the constant lmbda (gauge field term).")
        ("lmbda_constant_therm,lmbdact", po::value(&config.param_spec.lmbda_therm), "Value of the constant lmbda for thermalization (gauge field term).")
        ("N_resamples,Nr", po::value(&config.sim_spec.N_resamples), "Number of bootstrap resamples.")
        ("custom_therm,cth", po::value(&config.sim_spec.custom_therm), "Whether custom thermalization is used (to probe hysteresis).")
        ("observables,obs", po::value(&config.sim_spec.observables)->multitoken(), "Observables that are measured.") // can take multiple observables at once
        ("seed,s", po::value(&config.sim_spec.seed), "Seed for the pseudorandom number generator.")
        ("basis,bas", po::value(&config.lat_spec.basis), "Spin basis (\"x\" or \"z\").")
        ("output_directory,outdir", po::value(&config.out_spec.path_out), "Directory where the output is stored.")
        ("folder_name,fn", po::value(&config.out_spec.folder_name), "Name of the output subfolder.")
        ("folder_names,fns", po::value(&config.out_spec.folder_names)->multitoken(), "Directories where hysteresis output is stored.")
        ("snapshots,snap", po::value(&config.out_spec.save_snapshots), "Whether snapshots should be saved.")
        ("full_time_series,fts", po::value(&config.out_spec.full_time_series), "Whether full time series should be saved.")
        ("process_index,procid", po::value(&process_index), "Identifier of process (for debugging).")
        ("lattice_type,lat", po::value(&config.lat_spec.lattice_type), "Type of lattice used.")
        ("system_size,L", po::value(&config.lat_spec.system_size), "System size of lattice (one coordinate).")
        ("boundaries,bound", po::value(&config.lat_spec.boundaries), "Boundary condition of the lattice (periodic or open).")
        ("default_spin,dsp", po::value(&config.lat_spec.default_spin), "Default spin (electric field) for lattice initialization.");

        // With long-option disguise enabled, Boost otherwise treats "-h" as
        // an ambiguous prefix of the h-related options before resolving the
        // explicit short alias.
        for (int i = 1; i < argc; ++i) {
            if (std::string_view(argv[i]) == "-h") {
                std::cout << desc << '\n';
                return 0;
            }
        }

        po::variables_map vm;
        po::command_line_parser parser(argc, argv);
        parser.options(desc).style(
            po::command_line_style::default_style
            | po::command_line_style::allow_long_disguise
        );
        po::store(parser.run(), vm);
        po::notify(vm);

        if (vm.count("help") || argc == 1) {
            std::cout << desc << '\n';
            return 0;
        }
        if (!vm.count("simulation")) {
            throw std::invalid_argument("simulation must be specified");
        }

        if (simulation == "etc_sample") {
            paratoric::detail::validate_sample_config(config);
        } else if (simulation == "etc_hysteresis") {
            if (config.out_spec.folder_names.size() != config.param_spec.h_hys.size()) {
                throw std::invalid_argument(
                    "folder_names must match the hysteresis schedule length"
                );
            }
            auto config_to_validate = config;
            config_to_validate.out_spec.paths_out.resize(
                config.out_spec.folder_names.size()
            );
            paratoric::detail::validate_hysteresis_config(config_to_validate);
        } else if (simulation == "etc_thermalization") {
            paratoric::detail::validate_thermalization_config(config);
        } else {
            throw std::invalid_argument(
                "simulation must be 'etc_sample', 'etc_hysteresis', or 'etc_thermalization'"
            );
        }

        BOOST_LOG_TRIVIAL(info) << "Subprocess [ID=" << process_index 
                                << "] SPAWNED with [β=" << config.lat_spec.beta
                                << "], [μ=" << config.param_spec.mu
                                << "], [J=" << config.param_spec.J
                                << "], [h=" << config.param_spec.h
                                << "], [λ=" << config.param_spec.lmbda << "].";

        std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();

        if (simulation == "etc_sample") {
            auto io = paratoric::IO();
            io.etc_sample(config);
        } else if (simulation == "etc_hysteresis") {
            auto io = paratoric::IO();
            io.etc_hysteresis(config);
        } else if (simulation == "etc_thermalization") {
            auto io = paratoric::IO();
            io.etc_thermalization(config);
        }

        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();

        const auto dt = end - begin;
        const auto days = std::chrono::duration_cast<std::chrono::days>(dt);
        const std::chrono::hh_mm_ss<std::chrono::microseconds> hms{floor<std::chrono::microseconds>(dt - days)};

        BOOST_LOG_TRIVIAL(info)
        << std::format(
            "Subprocess [ID={}] COMPLETED in [{}-{:02}:{:02}:{:02}.{:06} h].",
            process_index,
            days.count(),
            hms.hours().count(),
            hms.minutes().count(),
            hms.seconds().count(),
            hms.subseconds().count());

        return 0;
    } catch (const std::exception& exc) {
        std::cerr << "Exception caught: " << exc.what() << std::endl;
        return 2;
    } 
}

File: python/cli/paratoric.py
# ParaToric - Continuous-time QMC for the extended toric code in the x/z-basis
# Copyright (C) 2022-2026  Simon Mathias Linsel, Lode Pollet

import argparse as ap
import logging
from math import pi
import multiprocessing
import sys

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

if not logger.handlers:
    formatter = logging.Formatter(fmt='%(asctime)s.%(msecs)03d - %(module)s - %(levelname)s - %(message)s',
                                  datefmt='%Y-%m-%d %H:%M:%S')

    console = logging.StreamHandler(stream=sys.stdout)
    console.setLevel(logging.INFO)
    console.flush = sys.stdout.flush
    console.setFormatter(formatter)
    logger.addHandler(console)


def cap_processes(args, limit):
    args['processes'] = min(args['processes'], limit)


def main(args):
    from job_handler import JobHandler

    lattice_params_list = ('lattice_type',
                           'system_size',
                           'boundaries',
                           'default_spin')

    lattice_params = {k: args[k] for k in lattice_params_list}

    if args['simulation'] == 'etc_T_sweep':
        cap_processes(args, args['T_steps'])
        job_handler = JobHandler()
        job_handler.etc_T_sweep(N_samples=args['N_samples'],
                                   N_thermalization=args['N_thermalization'],
                                   N_between_samples=args['N_between_samples'],
                                   T_lower=args['T_lower'],
                                   T_upper=args['T_upper'],
                                   T_steps=args['T_steps'],
                                   mu=args['mu_constant'],
                                   h=args['h_constant'],
                                   h_therm=args['h_constant_therm'],
                                   J=args['J_constant'],
                                   lmbda=args['lmbda_constant'],
                                   lmbda_therm=args['lmbda_constant_therm'],
                                   N_resamples=args['N_resamples'],
                                   custom_therm=args['custom_therm'],
                                   observables=args['observables'],
                                   seed=args['seed'],
                                   basis=args['basis'],
                                   save_snapshots=args['snapshots'],
                                   full_time_series=args['full_time_series'],
                                   processes=args['processes'],
                                   output_dir=args['output_directory'],
                                   **lattice_params)
        
    elif args['simulation'] == 'etc_hysteresis':
        cap_processes(args, 2)
        job_handler = JobHandler()
        job_handler.etc_hysteresis(N_samples=args['N_samples'],
                                      N_thermalization=args['N_thermalization'],
                                      N_between_samples=args['N_between_samples'],
                                      temperature=args['temperature'],
                                      mu=args['mu_constant'],
                                      h_hys=args['h_hysteresis'],
                                      J=args['J_constant'],
                                      lmbda_hys=args['lmbda_hysteresis'],
                                      N_resamples=args['N_resamples'],
                                      observables=args['observables'],
                                      seed=args['seed'],
                                      basis=args['basis'],
                                      save_snapshots=args['snapshots'],
                                      full_time_series=args['full_time_series'],
                                      processes=args['processes'],
                                      output_dir=args['output_directory'],
                                      **lattice_params)

    elif args['simulation'] == 'etc_h_sweep':
        cap_processes(args, args['h_steps'])
        job_handler = JobHandler()
        job_handler.etc_h_sweep(N_samples=args['N_samples'],
                                   N_thermalization=args['N_thermalization'],
                                   N_between_samples=args['N_between_samples'],
                                   temperature=args['temperature'],
                                   mu=args['mu_constant'],
                                   h_lower=args['h_lower'],
                                   h_upper=args['h_upper'],
                                   h_steps=args['h_steps'],
                                   h_therm=args['h_constant_therm'],
                                   J=args['J_constant'],
                                   lmbda=args['lmbda_constant'],
                                   lmbda_therm=args['lmbda_constant_therm'],
                                   N_resamples=args['N_resamples'],
                                   custom_therm=args['custom_therm'],
                                   observables=args['observables'],
                                   seed=args['seed'],
                                   basis=args['basis'],
                                   save_snapshots=args['snapshots'],
                                   full_time_series=args['full_time_series'],
                                   processes=args['processes'],
                                   output_dir=args['output_directory'],
                                   **lattice_params)

    elif args['simulation'] == 'etc_lmbda_sweep':
        cap_processes(args, args['lmbda_steps'])
        job_handler = JobHandler()
        job_handler.etc_lmbda_sweep(N_samples=args['N_samples'],
                                       N_thermalization=args['N_thermalization'],
                                       N_between_samples=args['N_between_samples'],
                                       temperature=args['temperature'],
                                       mu=args['mu_constant'],
                                       h=args['h_constant'],
                                       h_therm=args['h_constant_therm'],
                                       J=args['J_constant'],
                                       lmbda_lower=args['lmbda_lower'],
                                       lmbda_upper=args['lmbda_upper'],
                                       lmbda_steps=args['lmbda_steps'],
                                       lmbda_therm=args['lmbda_constant_therm'],
                                       N_resamples=args['N_resamples'],
                                       custom_therm=args['custom_therm'],
                                       observables=args['observables'],
                                       seed=args['seed'],
                                       basis=args['basis'],
                                       save_snapshots=args['snapshots'],
                                       full_time_series=args['full_time_series'],
                                       processes=args['processes'],
                                       output_dir=args['output_directory'],
                                       **lattice_params)
        
    elif args['simulation'] == 'etc_circle_sweep':
        cap_processes(args, args['Theta_steps'])
        job_handler = JobHandler()
        job_handler.etc_circle_sweep(N_samples=args['N_samples'],
                                        N_thermalization=args['N_thermalization'],
                                        N_between_samples=args['N_between_samples'],
                                        temperature=args['temperature'],
                                        mu=args['mu_constant'],
                                        h=args['h_constant'],
                                        J=args['J_constant'],
                                        lmbda=args['lmbda_constant'],
                                        radius=args['radius'],
                                        Theta_lower=args['Theta_lower'],
                                        Theta_upper=args['Theta_upper'],
                                        Theta_steps=args['Theta_steps'],
                                        N_resamples=args['N_resamples'],
                                        observables=args['observables'],
                                        seed=args['seed'],
                                        basis=args['basis'],
                                        save_snapshots=args['snapshots'],
                                        full_time_series=args['full_time_series'],
                                        processes=args['processes'],
                                        output_dir=args['output_directory'],
                                        **lattice_params)

    elif args['simulation'] == 'etc_thermalization':
        cap_processes(args, args['repetitions'])
        job_handler = JobHandler()
        job_handler.etc_thermalization(N_thermalization=args['N_thermalization'],
                                          repetitions=args['repetitions'],
                                          temperature=args['temperature'],
                                          mu=args['mu_constant'],
                                          h=args['h_constant'],
                                          J=args['J_constant'],
                                          lmbda=args['lmbda_constant'],
                                          N_resamples=args['N_resamples'],
                                          observables=args['observables'],
                                          seed=args['seed'],
                                          basis=args['basis'],
                                          save_snapshots=args['snapshots'],
                                          processes=args['processes'],
                                          output_dir=args['output_directory'],
                                          **lattice_params)
    
    else:
        raise ValueError(f'The simulation "{args["simulation"]}" does not exist.')


if __name__ == '__main__':
    parser = ap.ArgumentParser(description='Run extended toric code QMC simulation', formatter_class=ap.ArgumentDefaultsHelpFormatter)

    # Monte Carlo parameters
    mc_group = parser.add_argument_group('Simulation parameters')
    mc_group.add_argument('-sim', '--simulation',
                          help='Simulation you want to run.',
                          type=str, default='etc_T_sweep', 
                          choices=['etc_T_sweep', 
                                   'etc_hysteresis',
                                   'etc_h_sweep', 
                                   'etc_lmbda_sweep',
                                   'etc_circle_sweep',
                                   'etc_thermalization'])
    mc_group.add_argument('-Ns', '--N_samples', 
                          help='Number of snapshots of the simulation.', 
                          type=int, default=1000)
    mc_group.add_argument('-Nth', '--N_thermalization', 
                          help='Number of thermalization steps.', 
                          type=int, default=2000)
    mc_group.add_argument('-Nbs', '--N_between_samples', 
                          help='Number of steps between samples.', 
                          type=int, default=100)
    mc_group.add_argument('-reps', '--repetitions',
                          help='Number of MC simulations to average over.',
                          type=int, default=10)
    mc_group.add_argument('-T', '--temperature',
                          help='Temperature.',
                          type=float, default=1.)
    mc_group.add_argument('-Tl', '--T_lower',
                          help='Lower limit of temperature.',
                          type=float, default=1.)
    mc_group.add_argument('-Tu', '--T_upper',
                          help='Upper limit of temperature.',
                          type=float, default=2.)
    mc_group.add_argument('-Ts', '--T_steps',
                          help='Number of temperature steps.',
                          type=int, default=15)
    mc_group.add_argument('-muc', '--mu_constant',
                          help='Value of the constant mu (star term).',
                          type=float, default=1.)
    mc_group.add_argument('-Jc', '--J_constant', 
                          help='Value of the constant J (plaquette term).', 
                          type=float, default=1.)
    mc_group.add_argument('-hc', '--h_constant', 
                          help='Value of the constant h (electric field term).', 
                          type=float, default=0.)
    mc_group.add_argument('-hhys', '--h_hysteresis', nargs='+',
                          help='Hysteresis value of the constant h (electric field term).', 
                          type=float, default=[0.])
    mc_group.add_argument('-hct', '--h_constant_therm', 
                          help='Value of the constant h for thermalization (electric field term).', 
                          type=float, default=0.)
    mc_group.add_argument('-hl', '--h_lower',
                          help='Lower limit of h (electric field term).',
                          type=float, default=1.)
    mc_group.add_argument('-hu', '--h_upper',
                          help='Upper limit of h (electric field term).',
                          type=float, default=2.)
    mc_group.add_argument('-hs', '--h_steps',
                          help='Number of h steps (electric field term).',
                          type=int, default=15)
    mc_group.add_argument('-lmbdac', '--lmbda_constant', 
                          help='Value of the constant lmbda (gauge field term).', 
                          type=float, default=0.)
    mc_group.add_argument('-lmbdahys', '--lmbda_hysteresis', nargs='+',
                          help='Hysteresis value of the constant lmbda (gauge field term).', 
                          type=float, default=[0.])
    mc_group.add_argument('-lmbdact', '--lmbda_constant_therm',
                          help='Value of the constant lmbda for thermalization (gauge field term).',
                          type=float, default=0.)
    mc_group.add_argument('-lmbdal', '--lmbda_lower',
                          help='Lower limit of lmbda (gauge field term).',
                          type=float, default=1.)
    mc_group.add_argument('-lmbdau', '--lmbda_upper',
                          help='Upper limit of lmbda (gauge field term).',
                          type=float, default=2.)
    mc_group.add_argument('-lmbdas', '--lmbda_steps',
                          help='Number of lmbda steps (gauge field term).',
                          type=int, default=15)
    mc_group.add_argument('-rad', '--radius',
                          help='Radius of circle sweep in lmbda-h-space.',
                          type=float, default=0.5)
    mc_group.add_argument('-Thl', '--Theta_lower',
                          help='Lower limit of Theta.',
                          type=float, default=0.)
    mc_group.add_argument('-Thu', '--Theta_upper',
                          help='Upper limit of Theta.',
                          type=float, default=2*pi)
    mc_group.add_argument('-Ths', '--Theta_steps',
                          help='Number of Theta steps.',
                          type=int, default=15)
    mc_group.add_argument('-Nr', '--N_resamples', 
                          help='Number of bootstrap resamples.', 
                          type=int, default=1000)
    mc_group.add_argument('-cth', '--custom_therm', 
                          help='Whether custom thermalization is used (to probe hysteresis).', 
                          type=int, default=0, choices=[0, 1])
    mc_group.add_argument('-obs', '--observables', nargs='+',
                          help='Observables that are measured.',
                          type=str, default=['energy'])
    mc_group.add_argument('-s', '--seed', 
                          help='Seed for the pseudorandom number generator.', 
                          type=int, default=0)
    mc_group.add_argument('-bas', '--basis',
                          help='Spin basis (\"x\" or \"z\").',
                          type=str, default='x')
    mc_group.add_argument('-outdir', '--output_directory',
                          help='Directory where the output is stored.',
                          type=str, default=None)
    mc_group.add_argument('-snap', '--snapshots',
                          help='Whether snapshots should be saved.',
                          type=int, default=0, choices=[0, 1])
    mc_group.add_argument('-fts', '--full_time_series',
                          help='Whether full time series should be saved.',
                          type=int, default=0, choices=[0, 1])
    mc_group.add_argument('-proc', '--processes',
                          help='Number of processes. "0" will use all available cores. Negative numbers "-x" will use all available cores minus x.',
                          type=int, default=-4)

    # Lattice parameters
    lattice_group = parser.add_argument_group('Lattice parameters')
    lattice_group.add_argument('-lat', '--lattice_type', 
                               help='Type of lattice used.', 
                               type=str, default='square', choices=['square', 'cubic', 'honeycomb', 'triangular'])
    lattice_group.add_argument('-L', '--system_size',
                               help='System size of lattice (one coordinate).',
                               type=int, default=5)
    lattice_group.add_argument('-bound', '--boundaries', 
                               help='Boundary conditions.', 
                               type=str, default='periodic', choices=['periodic', 'open'])
    lattice_group.add_argument('-dsp', '--default_spin',
                               help='Default spin (electric field) for lattice initialization.',
                               type=int, default=1, choices=[-1, 1])

    args = vars(parser.parse_args())

    args['custom_therm'] = bool(args['custom_therm'])
    args['snapshots'] = bool(args['snapshots'])
    args['full_time_series'] = bool(args['full_time_series'])

    logical_cpu_core_count = multiprocessing.cpu_count()

    if args['processes'] == 0: 
        # Use all available cores when args['processes'] == 0
        args['processes'] = logical_cpu_core_count 
    elif logical_cpu_core_count < args['processes']:
        # Never use more processes than cores
        args['processes'] = logical_cpu_core_count
    elif args['processes'] < 0:
        # Use all available cores minus |args['processes']|
        if (logical_cpu_core_count + args['processes']) > 0:
            args['processes'] = logical_cpu_core_count + args['processes']
        else:
            args['processes'] = 1

    logger.info(f'Potentially use {args["processes"]} out of {logical_cpu_core_count} logical CPU cores.')

    main(args)
